### PR TITLE
ci: Update GitHub Actions to latest releases

### PR DESCRIPTION
**Pull request by matthewfeickert**
* Update GitHub Actions to latest releases.
* ~Enable weekly dependabot updates of GitHub Actions.~

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@alexander-held this is ready for review.

File: .github/dependabot.yml
Comment (alexander-held) [inline .github/dependabot.yml:12]: Hi @Sinclert, you mentioned in https://github.com/madminer-tool/madminer/pull/469 that the configuration ignoring major updates was for security purposes. Could you please explain this consideration? I would like to understand it to ensure that removing that here would not have any adverse effects.

Comment (matthewfeickert) [reply .github/dependabot.yml:12]: Given that dependabot opens PRs and there is no automerge, there isn't really a security issue as they need to pass CI before they could ever get merged in.

This rule is also keeping all of the official GitHub Actions from GitHub (anything under `actions/`) from ever getting updated.

Comment (Sinclert) [reply .github/dependabot.yml:12]: Hi both 👋🏻 

The reason to ignore major GHA updates (same case as in Python dependencies), is that major releases are used for compatibility breaking changes, according to [SemVer](https://semver.org). I usually activate Dependabot in repositories with the goal of keeping up to date with every **under-the-hood** change to the dependencies "APIs" I already hook my project to, whether that is a set of GHA Action expose parameters, or a set of public classes / functions in a Python package. Usually this is limited to improvements in performance ⚡ and security 🔒 

Any modification to the dependencies **public** API must be treated with extra-care, so a manual revision + syntax modification may be required. That is why I see no benefit in turning on their "_automatic version bumping_" PRs.

Comment (Sinclert) [reply .github/dependabot.yml:12]: I believe the reason why `pypa/gh-action-pypi-publish` version was not bumped in a Dependabot PR from `1.4.2` to `1.5.1` is that, due to the lack of development, Dependabot got deactivated a few months ago 🤔

Comment (matthewfeickert) [reply .github/dependabot.yml:12]: > The reason to ignore major GHA updates (same case as in Python dependencies), is that major releases are used for compatibility breaking changes, according to [SemVer](https://semver.org).

I appreciate the importance of SemVer, and this would make sense if they were doing something like (going with the Python analogy) deploying an application without pinning everything (bad idea) and were just putting an upper limit on the major releases, but that's not really the case. Keeping a GitHub Action at `@v2` is in Python parlance the same as using `~=2.0` so there's already no promise of SemVer stability here anyway as we know that [SemVer isn't a promise](https://hynek.me/articles/semver-will-not-save-you/). For the majority of GHA in the repo (all of them except `pypa/gh-action-pypi-publish`) having Dependabot avoiding major releases invalidates the Dependabot update check and will require manually updating them. That's an option, but that also means that someone is now responsible for manually checking whenever there is a release and that's not a realistic use of time and you might as well just remove the Dependabot check at that point.

> Any modification to the dependencies **public** API must be treated with extra-care, so a manual revision + syntax modification may be required. That is why I see no benefit in turning on their "_automatic version bumping_" PRs.

GitHub Actions aren't tools that can be installed and inspected in the abstract (there's no equivalent to making a fresh virtual environment on your local machine and then unpinning your Python dependencies and see if anything breaks) so there's no way way to know in advance if a new GHA release is going to cause problems beyond reading the release notes and running the action. If the Dependabot PR passes CI then you know nothing has been deprecated that will break your workflow, and then you can read the release notes to know if there have been any API additions that you might want to take advantage of. How would you go about validating a new major release of a GHA without running it?

If GitHub Actions SemVer is super important (`pypa/gh-action-pypi-publish` takes this seriously which is why they make suggestions down to the patch level) then none of these actions should be just at the `@vmajor` but they should be much more tightly restricted. For all of the `actions/` GHA that GitHub maintains this really isn't an issue so that luckily isn't needed.

Comment (Sinclert) [reply .github/dependabot.yml:12]: I have seen the [SemVer isn't a promise](https://hynek.me/articles/semver-will-not-save-you/) argument been used before, and while true, it does not change the fact that if a package (in this case a GHA Action) choose to use it, it can is a declaration of intent, that they will try at the best of their ability. There are very few things in software that are absolute promises, and yet, we keep operating.

In my experience, Dependabot PRs, due to their numbers and automatic nature, only scale well when they have limited scope. Otherwise, you will find yourself researching _what changed_ on `<Action X>` from v3 to v4, `<Action Y>` from v6 to v8, `<Action Z>` from v1 to v2, whenever Dependabot decides to trigger those PRs creation. On the other hand, it becomes a no-brainer to accept an `<Action X>` bump from v3.0.0 to v3.1.0, or an `<Action Y>` bump from v6.2.0 to 6.2.1, allowing developers to bump major versions **at their own pace, and only if they deemed it necessary**.

---

Removing the `ignore` code block on Dependabot's `github-actions` section will open the door for `pypa/gh-action-pypi-publish` to have their major version automatically bumped in Dependabot PR's, which is undesirable.

If you are wondering about _who_ would maintain GitHub Actions major version bumps, do not worry, I will. I haven't been doing it so far as I thought this project was dead.

Comment (matthewfeickert) [reply .github/dependabot.yml:12]: > Removing the `ignore` code block on Dependabot's `github-actions` section will open the door for `pypa/gh-action-pypi-publish` to have their major version automatically bumped in Dependabot PR's, which is undesirable.

I don't understand why you say this. Unless you are turning on automerge no such thing can happen without maintainers merging a dependabot PR.

This is slowing down an easy accept PR though so I'll move this all to an Issue and remove the dependabot changes from this PR.

**Review by alexander-held — APPROVED** [review]
Thanks! This looks good to me, and we can figure out the rest in #514.

**Review by Sinclert — APPROVED** [review]

**Pull request merged**
